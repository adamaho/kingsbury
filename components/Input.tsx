import * as React from 'react';
import * as classNames from 'classnames';
import styled from '../styles/theme';

import {
  ITheme
} from '../styles/theme';

interface IInputProps {
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  className?: string;
  placeholder?: string;
  disabled: boolean;
  theme: ITheme;
}

class InputComponent extends React.Component<IInputProps> {

  static defaultProps = {
    disabled: false
  }

  onInputChange  = (e: React.ChangeEvent<HTMLInputElement>) => {
    const {
      onChange
    } = this.props;

    if (onChange) {
      onChange(e);
    }
  }

  render() {
    const {
      className,
      disabled,
      placeholder
    } = this.props;

    return (
      <input
        className={classNames('kb-input', className, {
          disabled
        })}
        placeholder={placeholder}
        onChange={this.onInputChange}
      />
    );
  }
}

export const Input = styled(InputComponent)`
  height: ${(props: IInputProps) => props.theme.input.height};
  width: 100%;
  font-family: ${(props: IInputProps) => props.theme.fontFamily};
  font-size: 16px;
  background-color: ${(props: IInputProps) => props.theme.input.background};
  color: ${(props: IInputProps) => props.theme.colors.white};

  border: ${(props: IInputProps) => props.theme.input.border};
  border-color: ${(props: IInputProps) => props.theme.input.borderColor};
  border-radius: ${(props: IInputProps) => props.theme.input.borderRadius};
  box-sizing: border-box;

  line-height: 38px;
  padding: 10px 0px 5px 10px;

  &::placeholder {
    color: ${(props: IInputProps) => props.theme.input.placeholderColor};
  }

  &.disabled {
    cursor: not-allowed;
  }

  &:focus {
    border: ${(props: IInputProps) => `1px solid ${props.theme.colors.primary}`};
    outline: none;
  }
`;