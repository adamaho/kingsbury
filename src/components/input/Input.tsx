import * as React from 'react';

import styled, {
  css
} from 'styled-components';

import {
  theme
} from '../../theme';
import {SyntheticEvent} from "react";

export interface InputProps {

  /** Type of border for the input */
  borderType?: 'full' | 'bottom' | 'none';

  /** classname for the input */
  className?: string;

  /** Default value of the input */
  defaultValue?: string;

  /** Disabled state of the input */
  disabled?: boolean;

  /** Formik validation error */
  error?: string;

  /** Function that provides the error to use in custom component */
  errorComponent?: (error: string) => React.ReactNode;

  /** HTML input type attribute */
  htmlType?: 'text' | 'number' | 'date' | 'password';

  /** id of the input to be used with Formik */
  id?: string;

  /** Label of the input */
  label?: React.ReactNode;

  /** Function to handle blur event */
  onBlur?: React.EventHandler<SyntheticEvent>;

  /** Function to handle change event */
  onChange?: React.ChangeEventHandler<HTMLInputElement>;

  /** Function to handle focus event */
  onFocus?: React.EventHandler<SyntheticEvent>;

  /** Placeholder for the input */
  placeholder?: string;

  /** name of the input to be used with Formik */
  name?: string;

  /** Size of input */
  inputSize?: 'small' | 'large';

  /** Content to show inside the input to the left */
  inputPrefix?: React.ReactNode;

  /** Content to show inside the input to the right */
  inputSuffix?: React.ReactNode;

  /** Ref to be passed to the input */
  ref?: React.Ref<HTMLInputElement> | null;

  /** Disables typing in the input but keeps focus */
  readOnly?: boolean;

  /** Global theme in ThemeProvider */
  theme?: any;

  /** value of the input */
  value?: string;
}

const Container = styled.div`
  width: 100%;
`;

const Label = styled.label`
  font-size: 16px;
`;

const Error = styled.div<any>`
  color: ${(props) => props.theme.colors.danger};
`;

const InputFixContainer = styled.div`
  position: relative;
  width: 100%;
`;

const InputPrefix = styled.div<any>`
  display: flex;
  align-items: center;
  justify-content: center;

  position: absolute;
  height: ${(props) => props.inputSize === 'small' ?
    props.theme.input.heightSmall :
    props.theme.input.heightLarge
  };
  width: ${(props) => `${props.theme.input.prefixWidth}px`};
  left: ${(props) => props.theme.input.prefixLeft};
`;

const InputSuffix = styled.div<any>`
  display: flex;
  align-items: center;
  justify-content: center;

  position: absolute;
  height: ${(props) => props.inputSize === 'small' ?
    props.theme.input.heightSmall :
    props.theme.input.heightLarge
  };
  width: ${(props) => `${props.theme.input.suffixWidth}px`};
  right: ${(props) => props.theme.input.suffixRight};
`;

const StyledInput = styled.input<InputProps>`
  height: ${(props) => props.theme.input.heightSmall};
  font-size: ${(props) => props.theme.input.fontSizeSmall};

  ${(props) => props.inputSize === 'large' && css`
    height: ${props.theme.input.heightLarge};
    font-size: ${props.theme.input.fontSizeLarge};
  `};

  width: 100%;
  -webkit-appearance: none;

  font-family: inherit;

  background-color: ${(props) => props.theme.input.background};
  color: ${(props) => props.theme.input.color};

  ${(props) => props.error && css`
    border-color: ${props.theme.colors.danger};
  `};

  ${(props) => props.borderType === 'full' && css`
    padding: 10px 0px 10px 10px;
    border: ${props.theme.input.border};
    border-color: ${props.theme.input.borderColor};
    border-radius: ${props.theme.input.borderRadius};
    border-color: ${props.error ? props.theme.colors.danger : props.theme.input.borderColor};
  `};

  ${(props) => props.borderType === 'bottom' && css`
    padding: 10px 0px;
    border: none;
    border-radius: 0px;
    border-bottom: 1px solid;
    border-color: ${props.error ? props.theme.colors.danger : props.theme.input.borderColor};
  `};

  ${(props) => props.borderType === 'none' && css`
    padding: 10px 0px;
    border: none;
    color: ${props.error ? props.theme.colors.danger : props.theme.input.color};
  `};

  ${(props) => props.inputSuffix !== null && css`
    padding-right: ${props.theme.input.suffixWidth + 7}px
  `};

  ${(props) => props.inputPrefix !== null && css`
    padding-left: ${props.theme.input.prefixWidth + 7}px
  `};

  box-sizing: border-box;

  &:read-only {
    cursor: pointer;
  }

  &::placeholder {
    color: ${(props) => props.theme.input.placeholderColor};
  }

  &:disabled {
    cursor: not-allowed;
  }

  &:focus {
    border: ${(props) => `1px solid ${props.theme.colors.primary}`};
    ${(props) => props.borderType === 'bottom' && css`
      border: none;
      border-bottom: 1px solid ${props.theme.colors.primary};
      border-radius: 0px;
    `};

    ${(props) => props.borderType === 'none' && css`
      border: none;
    `};

    outline: none;
  }
`;

export const Input: React.FunctionComponent<InputProps> = React.forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  const {
    className,
    disabled,
    defaultValue,
    error,
    errorComponent,
    htmlType,
    id,
    label,
    name,
    onBlur,
    onChange,
    onFocus,
    placeholder,
    inputSize,
    inputPrefix,
    inputSuffix,
    borderType,
    theme: themeProp,
    readOnly,
    value
  } = props;

  return (
    <Container className={className}>
      {label && (
        <Label>{label}</Label>
      )}
        <InputFixContainer>
          {inputPrefix && (
            <InputPrefix
              inputSize={inputSize}
              theme={themeProp}
            >
              {inputPrefix}
            </InputPrefix>
          )}
          {inputSuffix && (
            <InputSuffix
              inputSize={inputSize}
              theme={themeProp}
            >
              {inputSuffix}
            </InputSuffix>
          )}
        </InputFixContainer>
        <StyledInput
          label={null}
          error={error}
          disabled={disabled}
          defaultValue={defaultValue}
          type={htmlType}
          id={id}
          name={name}
          onBlur={onBlur}
          onChange={onChange}
          onFocus={onFocus}
          placeholder={placeholder}
          borderType={borderType}
          inputSize={inputSize}
          inputSuffix={inputSuffix}
          inputPrefix={inputPrefix}
          ref={ref}
          readOnly={readOnly}
          theme={themeProp}
          value={value}
        />
      {(error && errorComponent) &&
        <Error theme={themeProp}>
          {errorComponent(error)}
        </Error>
      }
    </Container>
  );
});

Input.defaultProps = {
  borderType: 'full',
  className: '',
  disabled: false,
  error: '',
  errorComponent: (error) => <div>{error}</div>,
  htmlType: undefined,
  id: undefined,
  name: undefined,
  label: '',
  onBlur: undefined,
  onChange: undefined,
  onFocus: undefined,
  placeholder: '',
  inputSize: 'small',
  inputPrefix: null,
  inputSuffix: null,
  readOnly: false,
  theme,
  value: undefined
};
