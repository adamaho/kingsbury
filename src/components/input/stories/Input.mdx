import { Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { Title, SectionTitle } from '../../../../.storybook/components';
import { Input } from '../Input';

<Title
  title="Input"
  subtitle="Inputs allow users to input data."
/>

<SectionTitle>Simple</SectionTitle>

Many times a simple input is all that is needed.

<Preview>
  <Story id="components-input--simple"/>
</Preview>

<SectionTitle>Size</SectionTitle>

Change the size of the input by modifying the `inputSize` prop.

<Preview>
  <Story id="components-input--size"/>
</Preview>

<SectionTitle>Border</SectionTitle>

Modify how the `input` looks by changing the `borderType` prop;

<Preview>
  <Story id="components-input--border"/>
</Preview>

<SectionTitle>Label</SectionTitle>

Provide a `label` for the input to use in a form.

<Preview>
  <Story id="components-input--label"/>
</Preview>

<SectionTitle>Error</SectionTitle>

When using an `input` with form validation, provide the input with an `error` message and optional `errorComponent`.

<Preview>
  <Story id="components-input--error"/>
</Preview>

<SectionTitle>Affix</SectionTitle>

Add a prefix and suffix to an input by providing the props `inputPrefix` and `inputSuffix`.

<Preview>
  <Story id="components-input--affix"/>
</Preview>

<SectionTitle>Props</SectionTitle>
<Props of={Input} />
