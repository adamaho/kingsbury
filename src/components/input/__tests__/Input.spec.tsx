import * as React from 'react';
import * as enzyme from 'enzyme';
import * as renderer from 'react-test-renderer';

import Input from '..';

describe('Input', () => {
  test('it renders without children', () => {
    const tree = renderer.create(<Input />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders with placeholder', () => {
    const tree = renderer.create(<Input placeholder="placeholder" />).toJSON();
    expect(tree).toMatchSnapshot();
  });


  test('it calls onChange', () => {
    const onChangeMock = jest.fn();
    const component = enzyme.mount(
      <Input onChange={onChangeMock} />
    );

    component.find('input').simulate('change');
    expect(onChangeMock).toHaveBeenCalled();
  });
});
