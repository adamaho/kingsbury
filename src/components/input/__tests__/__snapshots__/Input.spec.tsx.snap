// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input it renders with placeholder 1`] = `
.c0 {
  font-size: 16px;
}

.c1 {
  height: props.theme.input.heightSmall;
  font-size: props.theme.input.fontSizeSmall;
  width: 100%;
  -webkit-appearance: none;
  font-family: inherit;
  background-color: transparent;
  color: #000000;
  padding: 10px 0px 10px 10px;
  border: 1px solid;
  border-color: #C4C4C4;
  border-radius: 7px;
  border-color: #C4C4C4;
  box-sizing: border-box;
  line-height: 38px;
}

.c1::-webkit-input-placeholder {
  color: #C4C4C4;
}

.c1::-moz-placeholder {
  color: #C4C4C4;
}

.c1:-ms-input-placeholder {
  color: #C4C4C4;
}

.c1::placeholder {
  color: #C4C4C4;
}

.c1:disabled {
  cursor: not-allowed;
}

.c1:focus {
  border: 1px solid #5C6AC4;
  outline: none;
}

<div
  className=""
>
  <label
    className="c0"
  >
    
    <input
      className="c1"
      disabled={false}
      label={null}
      onChange={[Function]}
      placeholder="placeholder"
    />
  </label>
  
</div>
`;

exports[`Input it renders without children 1`] = `
.c0 {
  font-size: 16px;
}

.c1 {
  height: props.theme.input.heightSmall;
  font-size: props.theme.input.fontSizeSmall;
  width: 100%;
  -webkit-appearance: none;
  font-family: inherit;
  background-color: transparent;
  color: #000000;
  padding: 10px 0px 10px 10px;
  border: 1px solid;
  border-color: #C4C4C4;
  border-radius: 7px;
  border-color: #C4C4C4;
  box-sizing: border-box;
  line-height: 38px;
}

.c1::-webkit-input-placeholder {
  color: #C4C4C4;
}

.c1::-moz-placeholder {
  color: #C4C4C4;
}

.c1:-ms-input-placeholder {
  color: #C4C4C4;
}

.c1::placeholder {
  color: #C4C4C4;
}

.c1:disabled {
  cursor: not-allowed;
}

.c1:focus {
  border: 1px solid #5C6AC4;
  outline: none;
}

<div
  className=""
>
  <label
    className="c0"
  >
    
    <input
      className="c1"
      disabled={false}
      label={null}
      onChange={[Function]}
      placeholder=""
    />
  </label>
  
</div>
`;
